docs/overview.rst

```diff
@@ -1,11 +1,410 @@
+
+    * Use :class:`IndexableDataset` to interface with indexable objects in
+      general.
+
+      - :class:`IndexableDataset` accepts an ``indexables`` argument, which is
+        expected to be a :class:`dict` mapping from source names to their
+        corresponding indexable objects.
+      - The state :meth:`IndexableDataset.open` returns is ``None``.
+      - Its :meth:`get_data` method expects a list of indices as its
+        ``request`` keyword argument.
+      - As such, it allows random access.
+
+The :class:`Dataset` class is responsible for interfacing with the data and
+handling data access requests. Subclasses of :class:`Dataset` specialize in
+certain types of data.
+
+Datasets contain one or more **sources** of data, such as an array of images,
+a list of labels, a dictionary specifying an ontology, etc. Each source in a
+dataset is identified by a unique name.
+
+IterableDataset
+^^^^^^^^^^^^^^^
+
+The simplest :class:`Dataset` subclass is :class:`IterableDataset`, which
+interfaces with iterable objects.
+
+It is created by passing a :class:`dict` mapping source names to their
+associated data and, optionally, a :class:`dict` mapping source names to tuples
+of axis labels.
+
+>>> from collections import OrderedDict
+>>> from fuel.datasets import IterableDataset
+>>> dataset = IterableDataset(
+...     iterables=OrderedDict([('features', features), ('targets', targets)]),
+...     axis_labels=OrderedDict([('features', ('batch', 'height', 'width')),
+...                              ('targets', ('batch', 'index'))]))
+
+We can ask the dataset what sources of data it provides by accessing its
+``sources`` attribute. We can also know which axes correspond to what by
+accessing its ``axis_labels`` attribute. It also has a ``num_examples`` property
+telling us the number of examples it contains.
+
+>>> print('Sources are {}.'.format(dataset.sources))
+Sources are ('features', 'targets').
+>>> print('Axis labels are {}.'.format(dataset.axis_labels))
+Axis labels are OrderedDict([('features', ('batch', 'height', 'width')), ('targets', ('batch', 'index'))]).
+>>> print('Dataset contains {} examples.'.format(dataset.num_examples))
+Dataset contains 8 examples.
+
+.. tip::
+
+   The source order of an :class:`IterableDataset` instance depends on the key
+   order of ``iterables``, which is nondeterministic for regular :class:`dict`
+   instances. We therefore recommend that you use
+   :class:`collections.OrderedDict` instances if the source order is important
+   to you.
+
+Datasets themselves are stateless objects (as opposed to, say, an open file
+handle, or an iterator object). In order to request data from the dataset, we
+need to ask it to instantiate some stateful object with which it will interact.
+This is done through the :meth:`Dataset.open` method:
+
+>>> state = dataset.open()
+>>> print(state.__class__.__name__)
+imap
+
+We see that in :class:`IterableDataset`'s case the state is an iterator object.
```

Comment:
Not really sure about my English here, but "We can see" sounds better to me.

Again, if `Dataset` is not abstract/an interface, this part seems generic enough to go in that section.